tsdb: Only initialise out-of-order fields when required

codesome opened the pull request:
cc @jesusvazquez

jesusvazquez reviewed (approved):
LGTM

Do we have any existing benchmarks to see the impact on memory allocation? I guess its not extrictly necessary but would be nice to know.

Also this is probably the first fields extracted from memSeries? Are there any plans to do the same to other fields?

I wonder if this would be a good usecase for generics... have a minimal memseries that could be extended with other fields such as the ooo fields ... of course if there are no more usecases then its probably not worth it but if the objective here is to have a smaller footprint of memseries in memory maybe it could work.

bboreham commented:
> the impact on memory allocation? 

This PR removes net 32 bytes from every series, and adds 48 bytes per series with OOO data (40 bytes rounded up to nearest [span size](https://go.dev/src/runtime/sizeclasses.go)). And then those figures are doubled for typical GC settings.

> Also this is probably the first fields extracted from memSeries? Are there any plans to do the same to other fields?

I took some out - #11280, #11288, #11296. Ganesh has proposed #11777.
There are not many fields remaining that we expect to be mostly unused.

bboreham commented on tsdb/head_append.go:
```diff
@@ -1086,7 +1086,11 @@ func (a *headAppender) Commit() (err error) {
+	if s.oooFields == nil {
+		s.oooFields = &memSeriesOOOFields{}
+	}
+	oooFields := s.oooFields
```
Is this variable needed?

bboreham commented on tsdb/head_append.go:
```diff
@@ -1376,30 +1380,31 @@ func (s *memSeries) cutNewHeadChunk(
 	ref := s.mmapCurrentOOOHeadChunk(chunkDiskMapper)
 
-	s.oooHeadChunk = &oooHeadChunk{
+	s.oooFields.oooHeadChunk = &oooHeadChunk{
```
Why is it ok to assume `s.oooFields` is non-nil here?  Does it need a comment?

bboreham commented on tsdb/head_read.go:
```diff
@@ -195,7 +195,7 @@ func (s *memSeries) headChunkID(pos int) chunks.HeadChunkID {
 // * pos == len(s.oooMmappedChunks) refers to s.oooHeadChunk
 func (s *memSeries) oooHeadChunkID(pos int) chunks.HeadChunkID {
-	return chunks.HeadChunkID(pos) + s.firstOOOChunkID
+	return chunks.HeadChunkID(pos) + s.oooFields.firstOOOChunkID
```
Why is it ok to assume s.oooFields is non-nil here? Does it need a comment?

bboreham commented on tsdb/head_read.go:
```diff
@@ -348,29 +348,30 @@ func (s *memSeries) chunk(id chunks.HeadChunkID, chunkDiskMapper *chunks.ChunkDi
 	// is len(s.mmappedChunks), it represents the next chunk, which is the head chunk.
-	ix := int(cid) - int(s.firstOOOChunkID)
-	if ix < 0 || ix > len(s.oooMmappedChunks) {
+	ix := int(cid) - int(oooFields.firstOOOChunkID)
```
Why is it ok to assume s.oooFields is non-nil here? Does it need a comment?

bboreham reviewed (commented):
Good work; some thoughts below.

codesome commented:
@bboreham addressed your comments. About assuming that the ooo variable is not nil: I added a comment to those functions that the caller must ensure that the ooo variable is not nil.

bboreham commented on tsdb/head_wal.go:
```diff
@@ -499,7 +499,17 @@ func (h *Head) resetSeriesWithMMappedChunks(mSeries *memSeries, mmc, oooMmc []*m
+		if mSeries.ooo == nil {
+			mSeries.ooo = &memSeriesOOOFields{}
+		}
+		mSeries.ooo.oooHeadChunk = nil
+		mSeries.ooo.firstOOOChunkID = 0
+		mSeries.ooo.oooMmappedChunks = oooMmc
```
This could be `*mSeries.ooo = memSeriesOOOFields{ oooMmappedChunks: oooMmc }`, which is more resilient against fields being added.

codesome replied:
Done!

bboreham reviewed (approved):
LGTM!
